### DISCUSSION TITLE
Proxmox Installation: Is Docker an absolute requirement?

### DISCUSSION THREAD

**Discussion by dkerr64** [jasonacox/Powerwall-Dashboard#208] [General]
Is it an absolute requirement that Powerwall-dashboard must run with Docker?  I ask because I want to run this on a Proxmox server and it supports LXC-based containers, but not Docker.  The only recommended way to run Docker on Proxmox is to first install a VM and then run Docker inside of that.  But this seams kind of wasteful.

Thanks

**Comment by jasonacox** [comment]
Hi @dkerr64 !  The project is a collection of apps that can all run separately, either as LXC containers or standalone.  The use of docker-compose was to help make the setup easy, but with some manual work you can assemble the whole thing on your host.  If you do take on this adventure, can you please log all the steps needed so that it helps others following you?

Some tips for manual setup:
* pypowerwall - see https://github.com/jasonacox/pypowerwall/tree/main/proxy
* influxdb - https://docs.influxdata.com/influxdb/v1.8/introduction/install/
* telegraf - https://docs.influxdata.com/telegraf/v1.25/install/
* weather411 (optional) - https://github.com/jasonacox/Powerwall-Dashboard/tree/main/weather
*  grafana - https://grafana.com/docs/grafana/latest/setup-grafana/installation/

**Reply by dkerr64** [reply]
@jasonacox thanks for the explanation.  I'll add it to my project list and if I succeed then I'll certainly document the steps.  I can see the logic an advantage to isolating each part into its own docker container.  It's just a pity that my current preferred server platform (Proxmox) does not support Docker.

I have to say that I love what you have built here.  I've only just started using it and am impressed.  Thank you.

David

**Reply by jasonacox** [reply]
Thanks @dkerr64 !  🙏

**Comment by wreiske** [comment]
I run this dashboard using proxmox on a Zotac, with debian 11 as an LXC, and then docker running inside the LXC.

With that being said, you could try these steps to convert it into an LXC.

How to run a docker container image directly as an LXC in proxmox:


Only works on proxmox 7 and up, it needs Debian bullseye (11.0 and up)


Install skopeo, umoci, and jq packages:
```
apt install skopeo umoci jq
```

Create an LXC rootfs from the docker image:
```
lxc-create name -t oci -- --url docker://alpine:latest

Getting image source signatures

Copying blob 2408cc74d12b [--------------------------------------] 0.0b / 0.0b

Copying config a366738a18 done  

Writing manifest to image destination

Storing signatures

Unpacking the rootfs
```

Turn the rootfs into a template that can be used from proxmox UI and the pct create command:

```
cd /var/lib/lxc/name/rootfs/

tar -czvf /var/lib/vz/template/cache/my_docker_container.tar.gz ./
```


Use the UI to start the template, or you could automate it like this:

```
pct create 100 /var/lib/vz/template/cache/my_docker_container.tar.gz \ -description LXC -hostname pvecontainer01 -memory 1024 -nameserver 8.8.8.8 \ -net0 name=eth0,hwaddr=52:4A:5E:26:58:D8,ip=192.168.15.147/24,gw=192.168.15.1,bridge=vmbr0 \ -storage local -password changeme
```

I haven't tested this method with a docker compose project, so it may require some fiddling.. 

I would recommend just creating an LXC container with Debian with docker inside of it.

**Reply by jasonacox** [reply]
Thanks @wreiske !

**Comment by dkerr64** [comment]
@wreiske thanks for the pointer. I had not come across this ability to "convert" a docker container into an LXC container.

When I was thinking about this I was not thinking of taking the five separate applications wrapped in docker and running them in five separate LXC containers.  Rather I was thinking of creating a single LXC container and running all the components in that one container.  There are of course pros-and-cons to this.

The other approach I was thinking of is to run Docker inside an LXC container.  This apparently works (I got the Hello World docker container to run) and would mean that there continues to be some isolation between each of the components plus also isolation from the underlying host.  But not at the expense of running a whole operating system in a VM.

David.

**Comment by dkerr64** [comment]
So, the way I decided to do this was to simply create a LXC container on proxmox... making sure that the `keyctl` and `nesting` were enabled and then installing docker inside that.  I used Ubuntu 22.04 as the base and followed instructions [here](https://docs.docker.com/engine/install/ubuntu/) to install docker.  Then I used the `setup.sh` easy-install for Powerwall-Dashboard.  And it is working like a charm.

I don't know why proxmox discourage using docker inside LXC, it certainly is working fine for me with this app.  It is running pretty lean too, I generously allocated 4GB of memory to the LXC but it looks like I can reduce that to 1GB and still have plenty to spare.  Is there a recommended minimum configuration for Powerwall-Dashboard?

**Reply by jasonacox** [reply]
Fantastic!  Thanks for posting the details. It will help others who have a Proxmax that want to get the dashboard running. I'm going to update the description title to make it easier for others to find.

My installations on a RPI and Ubuntu host are both using less than 1GB RAM.
